text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+
+````
+    struct SomeStruct<T> {...}
+
+    impl SomeStruct<i32> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Similar for generic `impl` blocks, e.g. `SomeStruct<T>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl<T> SomeStruct<T> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside trait `impl` blocks would use, enclosed in angle brackets,
+   the name of the type, followed by `as`, followed by the trait name, and
+   then `::` followed by the name of the function, e.g.
+   `<SomeStruct as SomeTrait>::some_function` for
+
+````
+    impl SomeTrait for SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Closures are using the name of the surrounding function, followed by `::`
+   and the string `<closure/definition_line_number>`, where `definition_line_number` is the number
+   of the line where the closure is defined. As closures are not named, coming
+   up with a meaningful name for them automatically is not easily possible.
+
+ - Nested functions are using the same scheme as closures, but instead of the
+   string `closure` they use their name. The line number of the definition is
+   included to prevent ambiguity with multiple nested functions with the same
+   name defined inside the same function.
+
+   If a nested function defines another nested function, the names are
+   appended to each other and separated by `::`, e.g.
+   `<fn1/definition_line_number_of_fn1::fn2/definition_line_number_of_fn2>`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Any addition to the standard library will need to be maintained forever, so it
+is worth weighing the maintenance cost over the value added. Given that this
+is a feature that is considered useful in other languages (e.g. `__FUNC__` in
+C) and is widely used there, it seems to be a useful addition to Rust too.
+
+Also adding a new macro to the standard library will make it impossible to use
+a macro with the same name in user code, because macros are (as of now) not
+namespaced.
+
+# Alternatives
+[alternatives]: #alternatives
+
+There are no alternatives to implement this in user code at this point, it
+would have to be provided by the compiler and standard library.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+ - Different name for the macro
+   - `function` would be nicer but easily conflicts with existing code
+   - `fn` would be consistent with how functions are declared
+   - `function_path`, makes more sense if the `module_path` would be
+     prepended (see below)
+ - Should a different naming scheme be used?
+   - An alternative would be using the naming scheme of backtraces, but
+     that is implementation defined and more disconnected from the code
+ - Should `module_path` be prepended or not?
+   - Seems redundant and easy to get in any case via the other macro
+ - A macro or intrinsic?
```

Comment:
This is not an unresolved question, but rather an implementation detail. Macro is possibly the only way this functionality would get exposed in a stable manner.